Mobile/ShopLens/Droid/Activities/CameraActivity.cs

```diff
@@ -187,127 +176,113 @@ protected override void OnActivityResult(int requestCode, Result resultCode, Int
-                    
-                    tts.Speak(
-                        $"This is. {guess}",
-                        QueueMode.Flush,
-                        null,
-                        null);
+                    if (!textRecognizer.IsOperational)
+                    {
+                        System.Diagnostics.Debug.WriteLine("[Warning]: Text recognizer not operational.");
+                        return new RecognitionResult { Predictions = await classifyImageTask };
+                    }
                     
-                }
-            })
+                    var recognizeTextTask = GetTextFromImageAsync(image);
+
+                    await Task.WhenAll(classifyImageTask, recognizeTextTask);
+
+                    var ocrResult = recognizeTextTask.Result;
+                    var weightString = new RegexMetricWeightSubstringFinder().FindWeightSpecifier(ocrResult);
+
+                    var recognitionResult = new RecognitionResult
+                    {
+                        RawOcrResult = ocrResult,
+                        Predictions = classifyImageTask.Result,
+                        WeightSpecifier = weightString
+                    };
+                    return recognitionResult;
+                })
                 .ContinueWith(task =>
                 {
                     progressBar.Visibility = ViewStates.Gone;
                     if (task.IsFaulted)
                     {
                         System.Diagnostics.Debug.WriteLine(task.Exception);
+                        return;
                     }
-                    if (task.IsCompletedSuccessfully)
-                    {
-                        shoppingCartErrorDialog.Show();                        
-                    }
+
+                    var recognitionResult = task.Result;
+                    var thingsToSay = new List<string>
+                        {$"This is {recognitionResult.BestPrediction}", recognitionResult.WeightSpecifier};
+                    
+                    tts.Speak(
+                        string.Join(". ", thingsToSay),
+                        QueueMode.Flush,
+                        null,
+                        null);
+
+                    var ocrResult = recognitionResult.RawOcrResult;
+                    if (!string.IsNullOrEmpty(ocrResult))
+                        new MessageBarCreator(rootView, $"OCR: {ocrResult}").Show();
+
+                    GetShoppingCartAddItemDialog(recognitionResult).Show();
                 }, TaskScheduler.FromCurrentSynchronizationContext());
         }
 
-        public void AddToShoppingCart()
+        private void AddToShoppingCart(string guess)
         {
-            prefs.AddString(guess.First().ToString().ToUpper() + guess.Substring(1));
-            shoppingCartMessageBar.Show();
+            var preferences = new ActivityPreferences(this, prefsName);
+            preferences.AddString(guess.FirstCharToUpper());
```

Comment:
What? We actually needed that extension method somewhere else other that Shopping Cart?! 😃 nice!

Reply:
Yeah, Tomas didn't notice we have the method for that.